crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -0,0 +1,106 @@
+    coin::Coin,
+    error::ExecutionError,
+    id::UID,
+    messages::{Command, ProgrammableTransaction},
+    storage::{ChildObjectResolver, ParentSync, Storage},
+};
+
+use super::{context::*, types::*};
+
+pub fn execute<
+    E: fmt::Debug,
+    S: ResourceResolver<Error = E>
+        + ModuleResolver<Error = E>
+        + Storage
+        + ParentSync
+        + ChildObjectResolver,
+>(
+    vm: &MoveVM,
+    state_view: &mut S,
+    ctx: &mut TxContext,
+    gas_status: &mut GasStatus,
+    gas_coin: ObjectID,
+    pt: ProgrammableTransaction,
+) -> Result<(), ExecutionError> {
+    let ProgrammableTransaction { inputs, commands } = pt;
+    let mut context = ExecutionContext::new(vm, state_view, ctx, gas_status, gas_coin, inputs)?;
+    for command in commands {
+        execute_command(&mut context, command)?;
+    }
+    Ok(())
+}
+
+fn execute_command<
+    E: fmt::Debug,
+    S: ResourceResolver<Error = E>
+        + ModuleResolver<Error = E>
+        + Storage
+        + ParentSync
+        + ChildObjectResolver,
+>(
+    context: &mut ExecutionContext<E, S>,
+    command: Command,
+) -> Result<(), ExecutionError> {
+    let is_transfer_objects = matches!(command, Command::TransferObjects(_, _));
+    let results = match command {
+        Command::TransferObjects(objs, addr_arg) => {
+            let objs: Vec<ObjectValue> = context.take_args(objs)?;
+            let addr: SuiAddress = context.take_arg(addr_arg)?;
+            for obj in objs {
+                obj.ensure_public_transfer_eligible()?;
+                context.transfer_object(obj, addr)?;
+            }
+            vec![]
+        }
+        Command::SplitCoin(coin_arg, amount_arg) => {
+            let mut obj: ObjectValue = context.take_arg(coin_arg)?;
+            let ObjectContents::Coin(coin) = &mut obj.contents else {
+                panic!("not a coin")
+            };
+            let amount: u64 = context.take_arg(amount_arg)?;
+            let new_coin_id = context.fresh_id()?;
+            let new_coin = coin.split_coin(amount, UID::new(new_coin_id))?;
+            let coin_type = obj.type_.clone();
+            context.restore_arg(coin_arg, Value::Object(obj))?;
+            vec![Some(Value::Object(ObjectValue::coin(coin_type, new_coin)?))]
+        }
+        Command::MergeCoins(target_arg, coin_args) => {
+            let mut target: ObjectValue = context.take_arg(target_arg)?;
+            let ObjectContents::Coin(target_coin) = &mut target.contents else {
+                panic!("not a coin")
+            };
+            let coins: Vec<ObjectValue> = context.take_args(coin_args)?;
+            for coin in coins {
+                let ObjectContents::Coin(Coin { id, balance }) = coin.contents else {
+                    panic!("not a coin")
+                };
+                context.delete_id(*id.object_id())?;
+                let Some(new_value) = target_coin.balance.value().checked_add(balance.value())
```

Comment:
Is it possible to use the implementation of `merge_coin` that you adapted here?

Reply:
I'd rather not... only that version of merge doesn't destroy the coins